internal/controller/warehouses/warehouses_test.go

```diff
@@ -1002,3 +1007,148 @@ func TestShouldDiscoverArtifacts(t *testing.T) {
+					// Intentionally not setting any xFns because we should never reach them
+					// because we will exit before any reconciliation logic is executed.
+				}
+			},
+			req: ctrl.Request{
+				NamespacedName: types.NamespacedName{
+					Name:      "test-warehouse",
+					Namespace: "test-namespace",
+				},
+			},
+			assertions: func(t *testing.T, r ctrl.Result, err error) {
+				require.NoError(t, err)
+				require.True(t, r.IsZero(), "expected no further reconciliation after shard mismatch")
+			},
+		},
+		{
+			name: "Shard match",
+			reconciler: func() *reconciler {
+				return &reconciler{
+					labelChecker: controller.ResponsibleFor[kargoapi.Warehouse]{
+						ShardName:           "right-shard",
+						IsDefaultController: false,
+					},
+					client: fake.NewClientBuilder().
+						WithScheme(testScheme).
+						WithObjects(
+							&kargoapi.Warehouse{
+								ObjectMeta: metav1.ObjectMeta{
+									Name:      "test-warehouse",
+									Namespace: "test-namespace",
+									Labels: map[string]string{
+										kargoapi.LabelKeyShard: "right-shard",
+									},
+								},
+							},
+						).Build(),
+					cfg: ReconcilerConfig{
+						ShardName:                 "right-shard",
+						MinReconciliationInterval: 5 * time.Minute,
+					},
+					discoverArtifactsFn: func(
+						context.Context,
+						*kargoapi.Warehouse,
+					) (*kargoapi.DiscoveredArtifacts, error) {
+						return &kargoapi.DiscoveredArtifacts{}, nil
+					},
+					discoverCommitsFn: func(
+						context.Context,
+						string,
+						[]kargoapi.RepoSubscription,
+					) ([]kargoapi.GitDiscoveryResult, error) {
+						return []kargoapi.GitDiscoveryResult{}, nil
+					},
+					discoverImagesFn: func(
+						context.Context,
+						string,
+						[]kargoapi.RepoSubscription,
+					) ([]kargoapi.ImageDiscoveryResult, error) {
+						return []kargoapi.ImageDiscoveryResult{}, nil
+					},
+					discoverChartsFn: func(
+						context.Context,
+						string,
+						[]kargoapi.RepoSubscription,
+					) ([]kargoapi.ChartDiscoveryResult, error) {
+						return []kargoapi.ChartDiscoveryResult{}, nil
+					},
+					buildFreightFromLatestArtifactsFn: func(
+						string,
+						*kargoapi.DiscoveredArtifacts,
+					) (*kargoapi.Freight, error) {
+						return &kargoapi.Freight{}, nil
+					},
+					createFreightFn: func(
+						context.Context,
+						client.Object,
+						...client.CreateOption,
+					) error {
+						return nil
+					},
+					patchStatusFn: func(
+						context.Context,
+						*kargoapi.Warehouse,
+						func(*kargoapi.WarehouseStatus),
+					) error {
+						return nil
+					},
+				}
+			},
+			req: ctrl.Request{
+				NamespacedName: types.NamespacedName{
+					Name:      "test-warehouse",
+					Namespace: "test-namespace",
+				},
+			},
+			assertions: func(t *testing.T, r ctrl.Result, err error) {
+				require.NoError(t, err)
+				require.False(t, r.IsZero(), "expected further reconciliation after shard match")
+			},
+		},
```

Comment:
Since we only added the two test cases we really needed to cover the change, we might put a TODO comment here calling for more test cases.